src/java.base/share/classes/jdk/internal/misc/CDS.java

```diff
@@ -59,4 +78,125 @@
+               name.equals(DELEGATING_HOLDER_CLASS_NAME)  ||
+               name.equals(BASIC_FORMS_HOLDER_CLASS_NAME) ||
+               name.equals(INVOKERS_HOLDER_CLASS_NAME);
+    }
+
+    private static boolean isBasicTypeChar(char c) {
+         return "LIJFDV".indexOf(c) >= 0;
+    }
+
+    private static boolean isValidMethodType(String type) {
+        String[] typeParts = type.split("_");
+        // check return type (second part)
+        if (typeParts.length != 2 || typeParts[1].length() != 1
+                || !isBasicTypeChar(typeParts[1].charAt(0))) {
+            return false;
+        }
+        // first part
+        if (!isBasicTypeChar(typeParts[0].charAt(0))) {
+            return false;
+        }
+        for (int i = 1; i < typeParts[0].length(); i++) {
+            char c = typeParts[0].charAt(i);
+            if (!isBasicTypeChar(c)) {
+                if (!(c >= '0' && c <= '9')) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    // return null for invalid input
+    private static Stream<String>  validateInputLines(String[] lines) {
+        ArrayList<String> list = new ArrayList<String>(lines.length);
+        for (String s: lines) {
+            String line = s.trim();
+            if (!line.startsWith("[LF_RESOLVE]") && !line.startsWith("[SPECIES_RESOLVE]")) {
+                System.out.println("Wrong prefix: " + line);
+                return null;
+            }
+
+            String[] parts = line.split(" ");
+            boolean isLF = line.startsWith("[LF_RESOLVE]");
+
+            if (isLF) {
+                if (parts.length != 4) {
+                    System.out.println("Incorrecct number of items in the line: " + parts.length);
+                    System.out.println("line: " + line);
+                    return null;
+                }
+                if (!isValidHolderName(parts[1])) {
+                    System.out.println("Invalid holder class name: " + parts[1]);
+                    return null;
+                }
+                if (!isValidMethodType(parts[3])) {
+                    System.out.println("Invalid method type: " + parts[3]);
+                    return null;
+                }
+            } else {
+                if (parts.length != 2) {
+                   System.out.println("Incorrect number of items in the line: " + parts.length);
+                   return null;
+                }
+           }
+           list.add(line);
+      }
+      return list.stream();
+    }
+
+    /**
+     * called from vm to generate MethodHandle holder classes
+     * @return @code { Object[] } if holder classes can be generated.
+     * @param lines in format of LF_RESOLVE or SPECIES_RESOLVE output
+     */
+    private static Object[] generateLambdaFormHolderClasses(String[] lines) {
+        Objects.requireNonNull(lines);
+        try {
+            Stream<String> lineStream = validateInputLines(lines);
```

Comment:
It seems clearer to have `validateInputLines` do validation only and convert this line into:

```
  validateInputLines(lines);
  Stream<String> lineStream = Arrays.stream(lines);
```

Reply:
Somewhere in the testing framework, the line in ExtraClassList was added '\f' which needs trim, I did not dig into deep the  root cause so here just return a new List with ending white spaces trimmed instead. I will file a bug for it and when it fixed, we can do this way.